src/VisualStudio/Core/Def/Implementation/FindUsages/RichNavFilterScopes.cs

```diff
@@ -0,0 +1,190 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.ComponentModel.Composition;
+using Microsoft.Internal.VisualStudio.Shell.ErrorList;
+using Microsoft.VisualStudio.Shell.TableControl;
+using Microsoft.VisualStudio.Shell.TableManager;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.FindUsages
+{
+    internal static class RichNavOptions
+    {
+        internal const string RichNavAvailableOptionName = "RichNavAvailable";
+    }
+
+    [Export(typeof(IScopeFilterFactory))]
+    [TableManagerIdentifier("FindAllReferences*")]
+    [TableManagerIdentifier("FindResults*")]
+    [Replaces(PredefinedScopeFilterNames.EntireSolutionScopeFilter)]
+    [Replaces(PredefinedScopeFilterNames.AllItemsScopeFilter)]
+    [DeferCreation(OptionName = RichNavOptions.RichNavAvailableOptionName)] // This factory will not be loaded unless this option is set to Boolean true
+    [DefaultScope]
+    [Name(PredefinedScopeFilterNames.LoadedSolutionScopeFilter)]
+    [Order(Before = PredefinedScopeFilterNames.AllItemsScopeFilter)]
+    internal class LoadedSolutionScopeFilterFactory : IReplacingScopeFilterFactory
+    {
+        [ImportingConstructor]
+        public LoadedSolutionScopeFilterFactory()
+        {
+        }
+
+        public IErrorListFilterHandler CreateFilter(IWpfTableControl tableControl)
+        {
+            // We're only replacing an existing filter, and not creating a new one.
+            return null;
+        }
+
+        public IErrorListFilterHandler ReplaceFilter(IWpfTableControl tableControl, string filterIdentifier)
+        {
+            if (filterIdentifier == PredefinedScopeFilterNames.AllItemsScopeFilter)
+            {
+                return new LoadedSolutionFilterHandler("Loaded Items");
+            }
+            else if (filterIdentifier == PredefinedScopeFilterNames.EntireSolutionScopeFilter)
+            {
+                return new LoadedSolutionFilterHandler("Loaded Solution");
+            }
+            return null; // Don't replace
+        }
+    }
+
+    internal class LoadedSolutionFilterHandler : FilterHandlerBase
+    {
+        private readonly string displayName;
+        internal const int LoadedSolutionFilterHandlerFilterId = 20;
```

Comment:
are the actual constant values of these relevant?  or is this just for ordering?

Reply:
They are relevant for programmatic selection. It's a part of a very old and internal API.
These values must be different for each handler. I guarantee that values in 20s are free to use.

Reply:
great.  so these are magic, and i can mark them as such :)